File: scraper/utils/parsing.py
"""
Utility functions for parsing dates, times, and other common data.
"""

import re
from datetime import datetime
from typing import Optional, Tuple
from dateutil import parser as dateutil_parser


def parse_date(date_string: str) -> Optional[str]:
    """
    Parse a date string into ISO 8601 format (YYYY-MM-DD).

    Args:
        date_string: Date string in various formats

    Returns:
        ISO 8601 formatted date string or None if parsing fails
    """
    if not date_string:
        return None

    try:
        # Use dateutil parser for flexible date parsing
        parsed_date = dateutil_parser.parse(date_string, fuzzy=True)
        return parsed_date.date().isoformat()
    except (ValueError, TypeError):
        return None


def parse_date_range(date_range_string: str) -> Tuple[Optional[str], Optional[str]]:
    """
    Parse a date range string into start and end dates.

    Examples:
        "Nov 14-15, 2025" -> ("2025-11-14", "2025-11-15")
        "November 14 & 15, 2025" -> ("2025-11-14", "2025-11-15")
        "Feb 21 - Mar 29" -> ("2025-02-21", "2025-03-29")

    Args:
        date_range_string: String containing a date range

    Returns:
        Tuple of (start_date, end_date) in ISO format, or (None, None)
    """
    if not date_range_string:
        return None, None

    # Common separators for date ranges
    separators = [' to ', ' - ', ' – ', ' through ', '–', '-']

    for sep in separators:
        if sep in date_range_string:
            parts = date_range_string.split(sep, 1)
            if len(parts) == 2:
                start = parse_date(parts[0].strip())
                end = parse_date(parts[1].strip())
                return start, end

    # Single date
    single_date = parse_date(date_range_string)
    if single_date:
        return single_date, single_date

    return None, None


def extract_price_range(text: str) -> Optional[str]:
    """
    Extract price range from text.

    Examples:
        "Tickets $20-$65" -> "$20-$65"
        "Starting at $5" -> "$5+"
        "Price: $35" -> "$35"

    Args:
        text: Text potentially containing price information

    Returns:
        Normalized price range string or None
    """
    if not text:
        return None

    # Pattern for price ranges like "$20-$65" or "$20 - $65"
    range_pattern = r'\$\s*(\d+(?:\.\d{2})?)\s*[-–]\s*\$?\s*(\d+(?:\.\d{2})?)'
    range_match = re.search(range_pattern, text)
    if range_match:
        return f"${range_match.group(1)}-${range_match.group(2)}"

    # Pattern for "starting at" or minimum prices
    starting_pattern = r'(?:starting\s+at|from)\s*\$\s*(\d+(?:\.\d{2})?)'
    starting_match = re.search(starting_pattern, text, re.IGNORECASE)
    if starting_match:
        return f"${starting_match.group(1)}+"

    # Pattern for single price
    single_pattern = r'\$\s*(\d+(?:\.\d{2})?)'
    single_match = re.search(single_pattern, text)
    if single_match:
        return f"${single_match.group(1)}"

    return None


def clean_text(text: str) -> str:
    """
    Clean and normalize text content.

    Args:
        text: Raw text to clean

    Returns:
        Cleaned text
    """
    if not text:
        return ""

    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text)

    # Remove leading/trailing whitespace
    text = text.strip()

    # Remove HTML entities
    text = text.replace('&amp;', '&')
    text = text.replace('&nbsp;', ' ')
    text = text.replace('&quot;', '"')
    text = text.replace('&#8217;', "'")
    text = text.replace('&#8220;', '"')
    text = text.replace('&#8221;', '"')

    return text


def extract_runtime(text: str) -> Optional[str]:
    """
    Extract runtime/duration from text.

    Examples:
        "90 minutes" -> "90 minutes"
        "2 hours" -> "2 hours"
        "1hr 30min" -> "1 hour 30 minutes"

    Args:
        text: Text potentially containing runtime information

    Returns:
        Normalized runtime string or None
    """
    if not text:
        return None

    # Pattern for "X hours Y minutes"
    hours_minutes = r'(\d+)\s*(?:hr|hour)s?\s*(?:(\d+)\s*(?:min|minute)s?)?'
    match = re.search(hours_minutes, text, re.IGNORECASE)
    if match:
        hours = int(match.group(1))
        minutes = int(match.group(2)) if match.group(2) else 0
        if minutes:
            return f"{hours} hour{'s' if hours != 1 else ''} {minutes} minutes"
        return f"{hours} hour{'s' if hours != 1 else ''}"

    # Pattern for just minutes
    minutes_pattern = r'(\d+)\s*(?:min|minute)s?'
    match = re.search(minutes_pattern, text, re.IGNORECASE)
    if match:
        return f"{match.group(1)} minutes"

    return None


def normalize_url(url: str, base_url: str) -> str:
    """
    Normalize a URL to be absolute.

    Args:
        url: URL to normalize (may be relative)
        base_url: Base URL for resolving relative URLs

    Returns:
        Absolute URL
    """
    if not url:
        return ""

    # Already absolute
    if url.startswith('http://') or url.startswith('https://'):
        return url

    # Protocol-relative
    if url.startswith('//'):
        return f"https:{url}"

    # Relative to base
    base = base_url.rstrip('/')
    path = url.lstrip('/')
    return f"{base}/{path}"

File: scraper/src/utils/parsing.ts
/**
 * Parsing utilities for dates, prices, and text normalization
 */

/**
 * Parse a date string to ISO 8601 format (YYYY-MM-DD)
 * Handles various common date formats
 */
export function parseDate(dateString: string): string | null {
  if (!dateString || typeof dateString !== 'string') {
    return null;
  }

  try {
    // Clean the string
    const cleaned = dateString.trim();

    // Try to parse as Date
    const date = new Date(cleaned);

    // Check if valid date
    if (isNaN(date.getTime())) {
      return null;
    }

    // Return ISO format (YYYY-MM-DD)
    return date.toISOString().split('T')[0];
  } catch (error) {
    return null;
  }
}

/**
 * Extract price range from text
 * Matches patterns like: $20, $20-$65, $20+, $20 - $65
 */
export function extractPriceRange(text: string): string | null {
  if (!text || typeof text !== 'string') {
    return null;
  }

  // Pattern 1: $20-$65 or $20 - $65
  const rangePattern = /\$(\d+)\s*[-–]\s*\$?(\d+)/;
  const rangeMatch = text.match(rangePattern);
  if (rangeMatch) {
    return `$${rangeMatch[1]}-$${rangeMatch[2]}`;
  }

  // Pattern 2: $20+ or $20 and up
  const plusPattern = /\$(\d+)\+/;
  const plusMatch = text.match(plusPattern);
  if (plusMatch) {
    return `$${plusMatch[1]}+`;
  }

  // Pattern 3: Single price $20
  const singlePattern = /\$(\d+)/;
  const singleMatch = text.match(singlePattern);
  if (singleMatch) {
    return `$${singleMatch[1]}`;
  }

  return null;
}

/**
 * Normalize whitespace in text
 * Collapses multiple spaces, removes leading/trailing whitespace
 */
export function normalizeWhitespace(text: string): string {
  if (!text || typeof text !== 'string') {
    return '';
  }

  return text
    .replace(/\s+/g, ' ') // Collapse multiple spaces
    .replace(/\n\s*\n/g, '\n') // Collapse multiple newlines
    .trim();
}

/**
 * Clean HTML text content
 * Removes extra whitespace and normalizes text
 */
export function cleanText(text: string | null | undefined): string {
  if (!text) {
    return '';
  }

  return normalizeWhitespace(text)
    .replace(/\u00A0/g, ' ') // Replace non-breaking spaces
    .replace(/\u200B/g, '') // Remove zero-width spaces
    .trim();
}

/**
 * Extract time from text
 * Matches patterns like: 8:00 PM, 8PM, 20:00
 */
export function extractTime(text: string): string | null {
  if (!text || typeof text !== 'string') {
    return null;
  }

  // Pattern: 8:00 PM or 8 PM
  const time12Pattern = /(\d{1,2}):?(\d{2})?\s*(AM|PM)/i;
  const time12Match = text.match(time12Pattern);
  if (time12Match) {
    const hours = time12Match[1];
    const minutes = time12Match[2] || '00';
    const period = time12Match[3].toUpperCase();
    return `${hours}:${minutes} ${period}`;
  }

  // Pattern: 20:00 (24-hour)
  const time24Pattern = /(\d{2}):(\d{2})/;
  const time24Match = text.match(time24Pattern);
  if (time24Match) {
    return time24Match[0];
  }

  return null;
}

/**
 * Extract JavaScript array from HTML text
 * Handles bracket matching, string delimiters, and escape sequences
 * Similar to Python version but adapted for TypeScript
 */
export function extractJsArray(text: string, startPattern: string): any[] | null {
  if (!text || !startPattern) {
    return null;
  }

  try {
    // Find the start pattern
    const startIndex = text.indexOf(startPattern);
    if (startIndex === -1) {
      return null;
    }

    // Find the opening bracket after the start pattern
    const searchStart = startIndex + startPattern.length;
    const openBracketIndex = text.indexOf('[', searchStart);
    if (openBracketIndex === -1) {
      return null;
    }

    // Track bracket depth and string delimiters
    let depth = 0;
    let inString = false;
    let stringDelimiter: string | null = null;
    let escaped = false;
    let closeBracketIndex = -1;

    // Safety limit to prevent runaway parsing
    const maxLength = Math.min(text.length, openBracketIndex + 500000); // 500KB limit

    for (let i = openBracketIndex; i < maxLength; i++) {
      const char = text[i];

      // Handle escape sequences
      if (escaped) {
        escaped = false;
        continue;
      }

      if (char === '\\') {
        escaped = true;
        continue;
      }

      // Handle string delimiters
      if (!inString && (char === '"' || char === "'" || char === '`')) {
        inString = true;
        stringDelimiter = char;
        continue;
      }

      if (inString && char === stringDelimiter) {
        inString = false;
        stringDelimiter = null;
        continue;
      }

      // Only count brackets outside of strings
      if (!inString) {
        if (char === '[') {
          depth++;
        } else if (char === ']') {
          depth--;
          if (depth === 0) {
            closeBracketIndex = i;
            break;
          }
        }
      }
    }

    if (closeBracketIndex === -1) {
      console.error('Could not find closing bracket for array');
      return null;
    }

    // Extract the array string
    let arrayString = text.substring(openBracketIndex, closeBracketIndex + 1);

    // Clean JavaScript syntax to make it valid JSON
    arrayString = arrayString
      .replace(/\\'/g, "'") // Fix escaped single quotes
      .replace(/,(\s*[}\]])/g, '$1'); // Remove trailing commas

    // Parse as JSON
    return JSON.parse(arrayString);
  } catch (error: any) {
    console.error(`Failed to extract JS array: ${error.message}`);
    return null;
  }
}

/**
 * Truncate text to a maximum length
 */
export function truncate(text: string, maxLength: number): string {
  if (!text || text.length <= maxLength) {
    return text;
  }

  return text.substring(0, maxLength - 3) + '...';
}

/**
 * Parse a date range from text
 * Returns [startDate, endDate] or [null, null] if parsing fails
 */
export function parseDateRange(text: string): [string | null, string | null] {
  if (!text || typeof text !== 'string') {
    return [null, null];
  }

  try {
    // Look for date range patterns like "Jan 1-15" or "Jan 1 - Feb 15"
    const rangePattern = /(\w+\s+\d+)\s*[-–]\s*(\w+\s+\d+|\d+)/;
    const match = text.match(rangePattern);

    if (match) {
      const startDateStr = match[1];
      let endDateStr = match[2];

      // If end date doesn't have month, use start month
      if (!/\w+\s+\d+/.test(endDateStr)) {
        const startMonth = startDateStr.split(' ')[0];
        endDateStr = `${startMonth} ${endDateStr}`;
      }

      const startDate = parseDate(startDateStr);
      const endDate = parseDate(endDateStr);

      return [startDate, endDate];
    }

    // Try to parse as single date
    const singleDate = parseDate(text);
    return [singleDate, null];
  } catch (error) {
    return [null, null];
  }
}
